### PR TITLE
Show message headers in fragment

### LINKED ISSUE #4504: Show header fields in separate screen ("show headers")
Currently we display header fields in the `MessageHeader` view inside `MessageViewFragment`. Instead we should open a new screen. This would be the beginning of the "show source" feature outlined here: https://github.com/k9mail/k-9/issues/1018#issuecomment-582539074

~Related: #3791 (headers are not displayed in the correct order)~

### PR DESCRIPTION
**Pull request by ByteHamster**
Opening the PR for early feedback.

Fixes weird scrolling behavior in the headers where sometimes text is selected. Closes #4504

- Will be converted to Kotlin as soon as it works as expected
- Already uses the name "view source" internally because that will likely be added later (see #1018)
- Open issues:
  - [x] Hide ActionBar menu when viewing source. Otherwise using the buttons creates a really strange backstack state. In general, K-9's backstack management looks pretty unusual to me...
  - [x] When pressing the back button, K-9 shows the message again but displays the sandwich menu in the ActionBar instead of an arrow. Clicking the sandwich button still navigates up, so only the icon is wrong.
  - [x] Test in split-screen mode, tablet mode, etc
  - [x] Is `messageLoaderHelperFactory.createForMessageView` actually what should be used here? It's not the message view fragment after all.

### CODE REVIEW

**Comment by cketti** [comment]
`MessageList` needs a serious overhaul. I think for now it's best to make the "view source" screen a separate Activity. That way you don't have to fight against `MessageList`'s back stack handling.

The "view source" screen should very much not use `MessageLoaderHelper*`. That, too, needs a serious overhaul. But it's also for loading messages for display, which does a lot of stuff "view source" does not need. Instead, create a `MessageRepository` class that fetches only the data we need (for now just the headers).

**Comment by ByteHamster** [comment]
Thanks! Should the screen use a TextView like before or a WebView? I think a WebView could be more efficient when displaying huge amounts of text like when adding the message source. We need to be more careful about XSS then, though.

**Comment by cketti** [comment]
I think we'll make showing the source of the body a separate fragment/screen. I don't have a good idea yet on how to deal with encoding and/or encryption. If you want to also work on that let's schedule a time to brainstorm a bit on how that feature should look like.

There are messages with quite a lot of header lines. I was thinking we could maybe use a `RecyclerView` to display them. But this is something we could change at a later point, too. For now the `TextView` should do.

**Comment by ByteHamster** [comment]
> If you want to also work on that let's schedule a time to brainstorm a bit on how that feature should look like.

I'm only interested in the headers, for now, to be honest. 

> I was thinking we could maybe use a RecyclerView to display them. 

A disadvantage of a RecyclerView would be that one cannot select text across multiple headers.

**Comment by cketti** [comment]
> I'm only interested in the headers, for now, to be honest.

That's totally fine :+1:

> A disadvantage of a RecyclerView would be that one cannot select text across multiple headers.

True. We'd have to do some extra work to allow selecting multiple list items and then offer a 'copy to clipboard' action. 

On the other hand, a RecyclerView would make it fairly easy to add actions for certain well-known headers. It probably also makes it easier to visually separate the header lines a little bit.

It's up to you. The RecyclerView approach is more work and only useful if more features are added later. The TextView approach is mostly copy & paste of existing code.

**Comment by ByteHamster** [comment]
> We'd have to do some extra work to allow selecting multiple list items and then offer a 'copy to clipboard' action.

Did that for one of my closed-source apps. Because of that, list items can only be selected as a whole. What if there are two types of users, some who copy parts of a single line and some who copy multiple lines? Anyway, I will go with a TextView for now. I was just annoyed by the text getting selected when scrolling through the headers, which is fixed when moving the TextView to its own fragment.

**Comment by ByteHamster** [comment]
I just force-pushed the Activity and another method to load the message with a `LocalMessageLoader` that only loads the message metadata. Not sure if that is what you meant when writing:

> Instead, create a MessageRepository class that fetches only the data we need

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageHeadersActivity.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.fsck.k9.activity
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageHeadersActivity.kt:1]: Please create a new package `com.fsck.k9.ui.messagesource`

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageHeadersActivity.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import android.view.MenuItem
+import com.fsck.k9.controller.MessageReference
+import com.fsck.k9.fragment.MessageHeadersFragment
+import com.fsck.k9.ui.R
+import com.fsck.k9.ui.base.K9Activity
+
+/**
+ * Temporary Activity used until the fragment can be displayed in the main activity.
+ */
+class MessageHeadersActivity : K9Activity() {
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageHeadersActivity.kt:13]: `AndroidManifest.xml` is missing an entry for this Activity.

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageList.kt [hunk 3]
```diff
@@ -879,9 +879,11 @@ open class MessageList :
         } else if (id == R.id.move_to_drafts) {
             messageViewFragment!!.onMoveToDrafts()
             return true
-        } else if (id == R.id.show_headers || id == R.id.hide_headers) {
-            messageViewFragment!!.onToggleAllHeadersView()
-            updateMenu()
+        } else if (id == R.id.show_headers) {
+            val intent = Intent(this, MessageHeadersActivity::class.java)
+            val reference = messageViewFragment!!.messageReference.toIdentityString()
+            intent.putExtra(MessageHeadersActivity.ARG_REFERENCE, reference)
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageList.kt:885]: This should be moved to `MessageHeadersActivity.kt`. Maybe as `createLaunchIntent()`.

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageRepository.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import androidx.loader.content.Loader;
+import com.fsck.k9.Account;
+import com.fsck.k9.Preferences;
+import com.fsck.k9.controller.MessageReference;
+import com.fsck.k9.controller.MessagingController;
+import com.fsck.k9.mailstore.LocalMessage;
+import com.fsck.k9.ui.message.LocalMessageLoader;
+import timber.log.Timber;
+
+public abstract class MessageRepository {
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/activity/MessageRepository.java:18]: Forget about trying to reuse any of the existing code. For rough guidance take a look at `FolderRepository`. For the loading/UI logic take a look at `ChangelogViewModel`.

If you don't want to deal with writing `MessageRepository` to access the database, let me know and I'll do it.

**Review by cketti — COMMENTED** [review]
You keep using this ancient programming language :smile:

**Review by ByteHamster — COMMENTED** [review]
> You keep using this ancient programming language

I find automatic type inference really hard to read - especially when reviewing code on GitHub where I cannot even hover over variables to see what type they are :smile: 

Just pushed the changes you requested (except the actual loading part)

**Comment by cketti** [comment]
> I find code with automatic type inference really hard to read - especially when reviewing code on GitHub where I cannot even hover over variables to see what type they are

I usually don't find this to be a problem. Most of the time you can read the code fine without having to know which type is used exactly.
Have you checked out the GitHub integration in IntelliJ? You can review pull requests with type hints enabled :)

**Comment by cketti** [comment]
I added my implementation of `MessageRepository` to this branch. Feel free to rewrite/rebase as necessary.

**Comment by ByteHamster** [comment]
> Have you checked out the GitHub integration in IntelliJ?

No, I haven't. Will check it out, thanks! :)

> I added my implementation of MessageRepository to this branch

Thanks! Looking at the code, I don't think I would have managed to do that :D Just updated my commit and pushed it.

**Comment by cketti** [comment]
@ByteHamster: any feedback on my latest changes?

**Comment by ByteHamster** [comment]
> any feedback on my latest changes?

I find it harder to read the code than before, to be honest, but I guess it's better design now :) Works fine on my device.

**Comment by cketti** [comment]
> I find it harder to read the code than before, to be honest

Anything in particular? I want to keep the general structure of Fragment + ViewModel + LiveData. But maybe cutting back on some Kotlin magic could make things easier to understand?

**Comment by ByteHamster** [comment]
I think it's just the fact that I'm not used to LiveData and the MVVM pattern (I already complained in #4222 :smile:). According to your explanation in https://github.com/k9mail/k-9/pull/4222#issuecomment-544764428, shouldn't `populateHeadersList` be located in the `ViewModel` class?

**Comment by cketti** [comment]
> According to your explanation in [#4222 (comment)](https://github.com/k9mail/k-9/pull/4222#issuecomment-544764428), shouldn't `populateHeadersList` be located in the `ViewModel` class?

In a way. The actual formatting is definitely a UI concern. But `MimeUtility.unfoldAndDecode()` shouldn't be called from the fragment. I left it there because I want to get rid of it eventually. Other clients don't decode headers when displaying the message source. I briefly removed the call and it messed up line breaks. I was too lazy to investigate the issue right then and left it as-is.

**Pull request merged**